src/S3/Parser/ValidateResponseChecksumResultMutator.php

```diff
@@ -0,0 +1,130 @@
+            return $result;
+        }
+
+        if (extension_loaded('awscrt')) {
+            $checksumPriority = ['CRC32C', 'CRC32', 'SHA1', 'SHA256'];
+        } else {
+            $checksumPriority = ['CRC32', 'SHA1', 'SHA256'];
+        }
+
+        $checksumsToCheck = array_intersect($responseAlgorithms, $checksumPriority);
+        $checksumValidationInfo = $this->validateChecksum($checksumsToCheck, $response);
+        if ($checksumValidationInfo['status'] == "SUCCEEDED") {
+            $result['ChecksumValidated'] = $checksumValidationInfo['checksum'];
+        } elseif ($checksumValidationInfo['status'] == "FAILED") {
+            //Ignore failed validations on GetObject if it's a multipart get which returned a full multipart object
+            if ($command->getName() === "GetObject"
+                && !empty($checksumValidationInfo['checksumHeaderValue'])
+            ) {
+                $headerValue = $checksumValidationInfo['checksumHeaderValue'];
+                $lastDashPos = strrpos($headerValue, '-');
+                $endOfChecksum = substr($headerValue, $lastDashPos + 1);
+                if (is_numeric($endOfChecksum)
+                    && intval($endOfChecksum) > 1
+                    && intval($endOfChecksum) < 10000) {
+                    return $result;
+                }
+            }
+
+            throw new S3Exception(
+                "Calculated response checksum did not match the expected value",
+                $command
+            );
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param $checksumPriority
+     * @param ResponseInterface $response
+     */
+    public function validateChecksum($checksumPriority, ResponseInterface $response): array
+    {
+        $checksumToValidate = $this->chooseChecksumHeaderToValidate(
+            $checksumPriority,
+            $response
+        );
+        $validationStatus = "SKIPPED";
+        $checksumHeaderValue = null;
+        if (!empty($checksumToValidate)) {
+            $checksumHeaderValue = $response->getHeader(
+                'x-amz-checksum-' . $checksumToValidate
+            );
+            if (isset($checksumHeaderValue)) {
+                $checksumHeaderValue = $checksumHeaderValue[0];
+                $calculatedChecksumValue = $this->getEncodedValue(
+                    $checksumToValidate,
+                    $response->getBody()
+                );
+                $validationStatus = $checksumHeaderValue == $calculatedChecksumValue
+                    ? "SUCCEEDED"
+                    : "FAILED";
+            }
+        }
+        return [
+            "status" => $validationStatus,
+            "checksum" => $checksumToValidate,
+            "checksumHeaderValue" => $checksumHeaderValue,
+        ];
+    }
+
+    /**
+     * @param $checksumPriority
+     * @param ResponseInterface $response
+     */
+    public function chooseChecksumHeaderToValidate(
```

Comment:
does this need to be public? same for the method above

Reply:
It is also public in the `ValidateResponseChecksumParser` class. It also has some testings depending on this method to be public.

Reply:
This hasn't been addressed